File: 82.remove-duplicates-from-sorted-list-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        dummy = ListNode(0)
        dummy.next = head
        node = dummy
        while node.next:
            next = node.next
            val = next.val
            if next.next and next.next.val == val:
                next = next.next.next
                while next and next.val == val:
                    next = next.next
                node.next = next
            else:
                node = node.next
        return dummy.next

File: 82.remove-duplicates-from-sorted-list-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        ListNode dummy(0);
        dummy.next = head;
        ListNode* node = &dummy;
        while (node->next) {
            ListNode* next = node->next;
            int val = next->val;
            if (next->next && next->next->val == val) {
                next = next->next->next;
                while (next && next->val == val) {
                    next = next->next;
                }
                node->next = next;
            } else {
                node = node->next;
            }
        }
        return dummy.next;
    }
};